contributors/design-proposals/storage/csi-migration.md

```diff
@@ -63,59 +102,88 @@ internal APIs.
-  default
+* Feature flags for Phase 1 and 2 enabled by default, per-plugin toggle on for
+  relevant cloud provider by default
 * CSI Drivers for migrated plugins available on related cloud provider cluster
   by default
 
 ## Milestones
 
 * Translation Library implemented in Kubernetes staging
-* Migration Shim for Provision, Attach, Detach, Mount, Unmount (including Inline Volumes)
-* Migration Shim for Resize, Block
+* Translation of volumes in volume controllers to support Provision, Attach,
+  Detach, Mount, Unmount (including Inline Volumes) using migrated CSI plugins.
+* Translation of volumes in volume controllers to support Resize, Block using
+  migrated CSI plugins.
 * CSI Driver lifecycle manager
 * GCE PD feature parity in CSI with in-tree implementation
 * AWS EBS feature parity in CSI with in-tree implementation
 * Cloud Driver feature parity in CSI with in-tree implementation
+* Skip ProbeVolumePlugins of migrated in-tree plugin code (Phase 2).
 
 ## Dependency Graph
 
 ![CSI Migration Dependency Diagram](csi-migration_dependencies.png?raw=true "CSI Migration Dependency Diagram")
 
 ## Feature Gating
 
-We will have an alpha feature gate for the whole feature that can turn the CSI
-migration on or off, when off all code paths should revert/stay with the in-tree
-plugins. We will also have individual flags for each driver so that admins can
-toggle them on or off.
+We will have two feature gates for the overall feature: CSIMigration and CSIMigrationInTreeOff
+corresponding to Phase 1 and 2 respectively. Additionally, plugin-specific feature flags
+(e.g. CSIMigrationGCE, CSIMigrationEBS) will determine whether a migration phase
+is enabled for a specific in-tree plugin. This allows administrators to enable a specific
+phase of migration functionality on the cluster as a whole as well as the flexibility
+to toggle migration functionality for each legacy in-tree plugin individually.
+
+With CSIMigration feature flag enabled on Kubernetes Controller Manager and Kubelet,
+several volume actions associated with in-tree plugins (that have plugin specific
+migration feature flags enabled) will be handled by CSI plugins that the in-tree
+plugins have migrated to. If the Kubelet on a cluster node does not have CSIMigration
+and plugin-specific migration feature flags enabled or running an old version of Kubelet
+before the CSI migration feature flags were introduced, the in-tree plugin code
+will continue to handle actions like attach/detach and mount/unmount of volumes on that node.
+
+During initialization, with CSIMigrationInTreeOff feature flag enabled, Kubernetes
+Controller Manager and Kubelet will skip invocation of ProbeVolumePlugins for migrated
+in-tree plugins (that have plugin-specific migration feature flags enabled).
+As a result, all nodes in the cluster are expected to have CSI plugins (that in-tree
+plugins have been migrated to) configured and installed and CSIMigration and
+plugin specific feature flags enabled.
```

Comment:
what happens if this expectation is violated? Might be worth mentioning that the post-migration code will still run and we'll just get errors - no fallback